src/device/device.cpp

```diff
@@ -39,32 +39,60 @@ namespace hw {
     
     /* ======================================================================= */
     /*  SETUP                                                                  */
-    /* ======================================================================= */   
-    device& get_device(const std::string device_descriptor) {
-        
-        struct s_devices {
-            std::map<std::string, std::unique_ptr<device>> registry;
-            s_devices() : registry() {
-                hw::core::register_all(registry);
-                #ifdef HAVE_PCSC
-                hw::ledger::register_all(registry); 
-                #endif
-           };
-        };
-        
-        static const s_devices devices;
+    /* ======================================================================= */
+
+    static std::unique_ptr<device_registry> registry;
+
+    device_registry::device_registry(){
+        hw::core::register_all(registry);
+        #ifdef HAVE_PCSC
+        hw::ledger::register_all(registry);
+        #endif
+    }
+
+    bool device_registry::register_device(const std::string & device_name, device * hw_device){
+        auto search = registry.find(device_name);
+        if (search != registry.end()){
+            return false;
+        }
+
+        registry.insert(std::make_pair(device_name, std::unique_ptr<device>(hw_device)));
+        return true;
+    }
+
+    device& device_registry::get_device(const std::string & device_descriptor){
+        // Device descriptor can contain further specs after first :
+        auto delim = device_descriptor.find(':');
+        auto device_descriptor_lookup = device_descriptor;
+        if (delim != std::string::npos) {
+            device_descriptor_lookup = device_descriptor.substr(0, delim);
+        }
```

Comment:
I wonder if adding a mutex for acessing the global would be better. It seems safe reading this anyway.